generic-lens/src/Data/Generics/Labels.hs

```diff
@@ -144,3 +163,35 @@ instance ( Applicative f, Choice p, Constructor name s t a b
 instance ( Applicative f, Choice p, Constructor name s t a b
          ) => IsLabelHelper 'ConstrType name p f s t a b where
   labelOutput = constructorPrism @name
+
+#if MIN_VERSION_base(4,18,0)
```

Comment:
Why do we actually need this? A comment would help! Looking at the issue, it seems like before 9.6 it is just the case that labels can't start with numbers. If that's all, then the code here is perfectly correct before 9.6, it just won't ever be able to trigger because no label of the correct form can be written in the surface syntax. Or am I missing something?

Reply:
Currently, the implementation uses eg [`UnconsSymbol`](https://hackage.haskell.org/package/base-4.19.1.0/docs/GHC-TypeLits.html#t:UnconsSymbol) which is only available since GHC 9.2 (see the ["`Char` kind" GHC proposal](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0387-char-kind.rst)), as the implementation before GHC 9.2 is very convoluted, see eg the [`symbols`](https://hackage.haskell.org/package/symbols) package.

As this feature is only interesting on GHCs >=9.6, I just made it conditional based on that. In particular, for some reason, the current definition of `DigitToNat` does not compile on GHC 9.2, only on >=9.4, but it didn't seem necessary to investigate that due to the fact that we only need GHC >=9.6 support.

Changed the condition to `#if MIN_VERSION_base(4,17,0)` and added a comment :+1:

Reply:
Aha, that makes sense. We could make the pre-9.6 case a TypeError so people get told why it doesn't work, but I don't think that's a big deal.

Reply:
Even then, you could leave `StartsWithDigit` and friends, and just have `ParseNat` do nothing pre 9.6. Not a big deal, it just can be nice to minimize the amount in the CPP.

Reply:
Good point, done :+1: